balancer/rls/picker.go

```diff
@@ -85,7 +86,13 @@ func (p *rlsPicker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
 	reqKeys := p.kbm.RLSKey(md, p.origEndpoint, info.FullMethodName)
 
 	p.lb.cacheMu.Lock()
-	defer p.lb.cacheMu.Unlock()
+	var pr balancer.PickResult
+	var err error
+	metricsCallback := func() {}
+	defer func() {
+		p.lb.cacheMu.Unlock()
+		metricsCallback()
+	}()
```

Comment:
What is the rationale behind changing the methods, `delegateToChildPoliciesLocked` and `useDefaultPickIfPossible` to return a callback that records the metrics instead of recording the metrics inline in those methods?

Reply:
We want to record this outside the cacheMu, since metrics recording can potentially be expensive. Doing it outside the cacheMu makes it not *block all other RPCs* which was Doug's main concern with doing inside cache mu.

Reply:
This is worth documenting with a comment.

Reply:
Done. Added comment "Record metrics without the cache mutex held, to prevent lock contention between concurrent RPC's and their Pick calls. Metrics Recording can potentially be expensive."